server/src/instant/runtime/routes.clj

```diff
@@ -253,27 +253,46 @@
                {:email email
                 :app-id app-id
                 :sub sub
-                :provider-id provider-id})]
-    (cond
-      (< 1 (count users))
-      (let [err (format "Got multiple app users for email=%s, sub=%s, provider-id=%s."
-                        email
-                        sub
-                        provider-id)]
-        (tracer/record-exception-span!
-         (Exception. err)
-         {:name "oauth/upsert-oauth-link!"
-          :escaping? false
-          :attributes {:email email
-                       :sub sub
-                       :user-ids (pr-str (map :app_user/id users))}})
-        nil)
-
-      (= 1 (count users))
-      (let [user (first users)]
+                :provider-id provider-id})
+
+        {:keys [user existing-oauth-link]}
+        (if (<= (count users) 1)
+          {:user (first users)
+           :existing-oauth-link nil}
+          (let [email-matches (when email
+                                (filter #(= email (:app_users/email %)) users))]
+            (when-not (= 1 (count email-matches))
+              (ex/throw-oauth-err! "Could not disambiguate between multiple users for this account."))
```

Comment:
Not: technically, I _don't_ think we can ever fall into this state. 

Given: 
1. We can't have more than 1 email in the database
2. We _must_ have had an email in our given param

This means: 
1. It can either find: 
 a. 1 item with the given sub 
 b. 2 items with the sub _and_ the email 

Now in all these cases we will disambiguate. 

Added an explicit place to throw an error here, so we can catch invariants early.